tools/ssimulacra2.cc

```diff
@@ -0,0 +1,435 @@
+    }
+  }
+
+  ssim *= 272.58355078216107;
+  ssim += -0.12953872587232876;
+
+  if (ssim > 0) {
+    ssim = 100.0 - 10.0 * pow(ssim, 0.6940375837127916);
+
+  } else {
+    ssim = 100.0;
+  }
+  return ssim;
+}
+
+void BminusY(jxl::Image3F& img) {
+  for (size_t y = 0; y < img.ysize(); ++y) {
+    const float* JXL_RESTRICT rowY = img.PlaneRow(1, y);
+    float* JXL_RESTRICT rowB = img.PlaneRow(2, y);
+    float* JXL_RESTRICT rowX = img.PlaneRow(0, y);
+    for (size_t x = 0; x < img.xsize(); ++x) {
+      rowB[x] += 1.0f - rowY[x];
+      rowX[x] += 0.5f;
+    }
+  }
+}
+Msssim ComputeSSIMULACRA2(jxl::Image3F& img1, jxl::Image3F& img2) {
+  Msssim msssim;
+
+  Image3F mul(img1.xsize(), img1.ysize());
+  Blur blur(img1.xsize(), img1.ysize());
+
+  for (int scale = 0; scale < kNumScales; scale++) {
+    if (img1.xsize() < 8 || img1.ysize() < 8) {
+      break;
+    }
+    if (scale) {
+      img1 = Downsample(img1, 2, 2);
+      img2 = Downsample(img2, 2, 2);
+    }
+    mul.ShrinkTo(img1.xsize(), img2.ysize());
+    blur.ShrinkTo(img1.xsize(), img2.ysize());
+
+    Multiply(img1, img1, &mul);
+    Image3F sigma1_sq = blur(mul);
+
+    Multiply(img2, img2, &mul);
+    Image3F sigma2_sq = blur(mul);
+
+    Multiply(img1, img2, &mul);
+    Image3F sigma12 = blur(mul);
+
+    Image3F mu1 = blur(img1);
+    Image3F mu2 = blur(img2);
+
+    MsssimScale sscale;
+    SSIMMap(mu1, mu2, sigma1_sq, sigma2_sq, sigma12, sscale.avg_ssim);
+    EdgeDiffMap(img1, mu1, img2, mu2, sscale.avg_edgediff);
+    msssim.scales.push_back(sscale);
+  }
+  return msssim;
+}
+
+namespace {
+
+int PrintUsage(char** argv) {
+  fprintf(stderr, "Usage: %s orig.png distorted.png\n", argv[0]);
+  return 1;
+}
+
+int Run(int argc, char** argv) {
+  if (argc != 3) return PrintUsage(argv);
+
+  jxl::CodecInOut io1;
+  jxl::CodecInOut io2;
+  JXL_CHECK(SetFromFile(argv[1], jxl::extras::ColorHints(), &io1));
+  JXL_CHECK(SetFromFile(argv[2], jxl::extras::ColorHints(), &io2));
+  JXL_CHECK(io1.TransformTo(jxl::ColorEncoding::LinearSRGB(io1.Main().IsGray()),
+                            jxl::GetJxlCms()));
+  JXL_CHECK(io2.TransformTo(jxl::ColorEncoding::LinearSRGB(io2.Main().IsGray()),
+                            jxl::GetJxlCms()));
+
+  if (io1.xsize() != io2.xsize() || io1.ysize() != io2.ysize()) {
+    fprintf(stderr, "Image size mismatch\n");
+    return 1;
+  }
+  if (io1.xsize() < 8 || io1.ysize() < 8) {
+    fprintf(stderr, "Minimum image size is 8x8 pixels\n");
+    return 1;
+  }
+  jxl::Image3F orig(io1.xsize(), io1.ysize());
+  jxl::ToXYB(io1.Main(), nullptr, &orig, jxl::GetJxlCms(), nullptr);
+  jxl::Image3F dist(io2.xsize(), io2.ysize());
+  jxl::ToXYB(io2.Main(), nullptr, &dist, jxl::GetJxlCms(), nullptr);
+  BminusY(orig);
+  BminusY(dist);
+
+  Msssim msssim = ComputeSSIMULACRA2(orig, dist);
+  printf("%.8f\n", msssim.Score());
+  return 0;
+}
+
+}  // namespace
+}  // namespace ssimulacra2
+
+int main(int argc, char** argv) { return ssimulacra2::Run(argc, argv); }
```

Comment:
main and run really should be in a different file if the intention is to allow using this as a library.

If not, then the .h file is not particularly useful :)

Reply:
Right. I guess having it as a library could be useful (say to be used in benchmark_xl), so I'll split it up.